### PR TITLE
ocicni: support for asynchronous CNI config changes

### PR DESCRIPTION
**Pull request by sameo**
We need to support cases where `InitCNI()` is called before
any CNI configuration files have been installed. This is
for example happening when deploying a k8s cluster with kubeadm.
kubeadm will start the DNS pod and it is left to the caller to
pick a network overlay and create the corresponding pods, that
will typically install a CNI configuration file first.
    
Here we address that issue by doing 2 things:
    
- Not returning an error when the default CNI config files
directory is empty.
- If it is empty, we start a monitoring thread (fsnotify based)
that will synchronize the network configuration when a CNI
file is installed there.

Together with #433 this allows us to reliably deploy a k8s 1.6 cluster with kubeadm.

### CODE REVIEW

**Comment by sameo** [comment]
@rajatchopra @mrunalp @runcom PTAL

**Comment by sameo** [comment]
@rajatchopra @mrunalp @dcbw
 
The reason why the `kube-dns` pod was not restarted upon `SetUpPod()` failure was because we were not cleaning our failed pod. Kubelet could then no longer get status for any pod and would not be able to decide if the pod had to be restarted or not.

I fixed that with PR #433. The networking changes needed are then fairly simpler than the previous PR.

**Comment by mrunalp** [comment]
@sameo lint is failing in the tests.

File: pkg/ocicni/ocicni.go [hunk 1]
```diff
@@ -163,6 +205,15 @@ func (plugin *cniNetworkPlugin) setDefaultNetwork(n *cniNetwork) {
 	plugin.defaultNetwork = n
 }
 
+func (plugin *cniNetworkPlugin) terminateMonitorNetDir(err error) {
+	plugin.Lock()
+	defer plugin.Unlock()
+	if !plugin.monitorNetDirDone {
```
Comment (rajatchopra) [inline pkg/ocicni/ocicni.go]: Why do we need this flag? We just have one monitor thread anyway.

Comment (sameo) [inline pkg/ocicni/ocicni.go]: @rajatchopra We don't need it, this is a copy/paste from my previous PR. This is fixed now.

File: pkg/ocicni/ocicni.go [hunk 2]
```diff
@@ -67,10 +109,13 @@ func probeNetworkPluginsWithVendorCNIDirPrefix(pluginDir string, cniDirs []strin
 		pluginDir:          pluginDir,
 		cniDirs:            cniDirs,
 		vendorCNIDirPrefix: vendorCNIDirPrefix,
+		monitorNetDirChan:  make(chan bool),
```
Comment (mikebrow) [inline pkg/ocicni/ocicni.go]: chan struct{}

File: pkg/ocicni/ocicni.go [hunk 3]
```diff
@@ -23,6 +23,8 @@ type cniNetworkPlugin struct {
 	pluginDir          string
 	cniDirs            []string
 	vendorCNIDirPrefix string
+
+	monitorNetDirChan chan bool
```
Comment (mikebrow) [inline pkg/ocicni/ocicni.go]: chan struct{}

**Comment by sameo** [comment]
@rajatchopra @mrunalp @runcom PTAL

**Comment by mrunalp** [comment]
LGTM. @rajatchopra Do you have any more comments?

**Comment by mrunalp** [comment]
Ran the k8s e2e tests with 115/121 :+1:

**Pull request merged**